draft-ietf-quic-tls.md

```diff
@@ -255,142 +205,206 @@ A simplified TLS 1.3 handshake with 0-RTT application data is shown in
+  QUIC providing a reliable stream abstraction to TLS.
 
-QUIC permits a client to send frames on streams starting from the first packet.
-The initial packet from a client contains a stream frame for stream 0 that
-contains the first TLS handshake messages from the client.  This allows the TLS
-handshake to start with the first packet that a client sends.
+* The TLS component provides a series of updates to the QUIC
+  component, including (a) new packet protection keys to install (b)
+  state changes such as handshake completion, the server certificate,
+  etc.
 
-QUIC packets are protected using a scheme that is specific to QUIC, see
-{{packet-protection}}.  Keys are exported from the TLS connection when they
-become available using a TLS exporter (see Section 7.5 of {{!TLS13}} and
-{{key-expansion}}).  After keys are exported from TLS, QUIC manages its own key
-schedule.
+{{schematic}} shows these interactions in more detail, with the QUIC
+packet protection being called out specially.
+
+~~~
++------------+                        +------------+
+|            |<- Handshake Messages ->|            |
+|            |<---- 0-RTT Keys -------|            |
+|            |<--- Handshake Keys-----|            |
+|   QUIC     |<---- 1-RTT Keys -------|    TLS     |
+|            |<--- Handshake Done ----|            |
++------------+                        +------------+
+ |         ^
+ | Protect | Protected
+ v         | Packet
++------------+
+|   QUIC     |
+|  Packet    |
+| Protection |
++------------+
+~~~
+{: #schematic title="QUIC and TLS Interactions"}
+
+Unlike TLS over TCP, QUIC applications which want to send data do not
+send it through TLS "application_data" records. Rather, they send it
+as QUIC STREAM frames which are then carried in QUIC packets.
+
+
+# Carrying TLS Messages {#carrying-tls}
+
+QUIC carries TLS handshake data in CRYPTO_HS frames, each of which
+consists of a contiguous block of handshake data (identified by an
+offset and length). Those frames are packaged into QUIC packets
+and encrypted under the current TLS encryption level.
+As with TLS over TCP, once TLS handshake data has
+been delivered to QUIC, it is QUIC's responsibility to deliver it
+reliably. Each chunk of data is associated with the then-current TLS
+sending keys, and if QUIC needs to retransmit that data, it MUST use
+the same keys even if TLS has already updated to newer keys.
+
+One important difference between TLS 1.3 records (used with TCP)
+and QUIC CRYPTO_HS frames is that in QUIC multiple frames may appear
+in the same QUIC packet as long as they are associated with the
+same encryption level. For instance, an implementation might
+bundle a Handshake message and an ACK for some Handshake
+data into the same packet.
+
+In general, the rules for which data can appear in packets of which
+encryption level are the same in QUIC as in TLS over TCP:
+
+- CRYPTO_HS frames MAY appear in packets of any encryption level.
```

Comment:
The text suggests that a CRYPTO_HS frame might appear in a 0-RTT packet. I think we need to forbid that.

IMO, the only handshake message that might make sense to show up in a 0-RTT packet is EndOfEarlyData. However, allowing that introduces unnecessary complexity.

We need to deliver the TLS messages to the TLS stack in the order they were issued.

A QUIC implementation can easily guarantee that plaintext messages (i.e. CH, SH) carried over the Initial packets will be delivered prior to handshake messages carried over the Handshake packets (e.g., EE, Certificate, Finished), because the latter cannot be decrypted without the former being delivered. The same argument goes for NST being delivered after ServerFinished.

If we send EOED using a Handshake packet, we will continue to have the in-order delivery guarantee.

OTOH, if we send EOED using a 0-RTT packet, QUIC stacks will be required to reassembly the input across multiple encryption levels, i.e., delay passing the payload of CRYPTO_HS frames found in Handshake packets until receiving a 0-RTT packet containing a EOED message. Note that a QUIC implementation cannot even tell if a CRYPTO_HS frame found in a 0-RTT packet contains a full EOED message, so we will be required to add a flag to a 0-RTT packet telling the peer that complete EOED has been delivered.

These requirements seem wired to me. Why not just send EOED using a Handshake packet?

Reply:
The other reason to send EOED using a Handshake packet would be to minimize the changes related to the security from TLS. In TLS, EOED is protected by the handshake traffic key. Sending EOED using a 0-RTT packet will mean that we will be changing the traffic key of the message to a different key.

Reply:
Please forgive me for being a complete idiot. The issues pointed out here is based on the assumption that the TLS stack will activate the handshake key before receiving EOED sent using a 0-RTT packet.

They are all non-issues, if the TLS stack delays the activation of the handshake key until it receives EOED (and that is what the stacks are doing now).